src/strategies/flex_window.py

```diff
@@ -84,170 +86,321 @@ def step(self, event_list=[]):
+                        p = util.clamp_power(power, sim_vehicle, cs)
+                        avg_power = sim_vehicle.battery.load(self.interval, p)[
+                            "avg_power"
+                        ]
+                    elif not charged_in_window and ts_info["window"]:
+                        # charging windows not sufficient, charge max during window
+                        p = util.clamp_power(ts_info["power"], sim_vehicle, cs)
+                        avg_power = sim_vehicle.battery.load(self.interval, p)[
+                            "avg_power"
+                        ]
+
+                    power_vec[ts_idx] = avg_power
+                    safe = sim_vehicle.get_delta_soc() <= 0
+                    if safe:
+                        # todo: why is this?
+                        power_vec[ts_idx + 1:] = [0] * (
+                            len(timesteps) - ts_idx - 1
+                        )
+                        break
 
-                charged_in_window = sim_vehicle.get_delta_soc() <= 0
+                if safe:
+                    max_power = power
+                else:
+                    min_power = power
+
+            # apply power
+            if gc.window:
+                p = (
+                    power
+                    if charged_in_window
+                    else gc.cur_max_power - gc.get_current_load()
+                )
+            else:
+                p = 0 if charged_in_window else power
+            p = util.clamp_power(p, vehicle, cs)
+            avg_power = vehicle.battery.load(self.interval, p)["avg_power"]
+            commands[cs_id] = gc.add_load(cs_id, avg_power)
+            cs.current_power += avg_power
 
-                if charged_in_window:
-                    # reset sim SoC
-                    sim_vehicle.battery.soc = vehicle.battery.soc
+            for ts_idx, ts_info in enumerate(timesteps):
+                ts_info["power"] -= power_vec[ts_idx]
 
-                min_power = 0
-                max_power = util.clamp_power(cs.max_power, sim_vehicle, cs)
-                old_soc = sim_vehicle.battery.soc
-                safe = False
-                power_vec = [0]*len(timesteps)
-                while (charged_in_window and not safe) or max_power - min_power > self.EPS:
-                    power = (min_power + max_power) / 2
-                    sim_vehicle.battery.soc = old_soc
+        return commands
 
-                    cur_time = self.current_time - self.interval
-                    for ts_idx, ts_info in enumerate(timesteps):
-                        cur_time += self.interval
-                        avg_power = 0
-                        if cur_time >= sim_vehicle.estimated_time_of_departure:
+    def distribute_balanced_batteries(self, timesteps):
+
+        gc = list(self.world_state.grid_connectors.values())[0]
+
+        batteries = [b for b in self.world_state.batteries.values()
+                     if b.parent is not None]
```

Comment:
do we have unconnected batteries? Even then, the check b.parent == gc_id would make more sense (would need the gc_id from somewhere, of course)